### PR TITLE
ARROW-16597: [Python][FlightRPC] Force server shutdown at interpreter exit

### PR DESCRIPTION
**Pull request by lidavidm**

### CODE REVIEW

File: python/pyarrow/_flight.pyx [hunk 1]
```diff
@@ -2516,6 +2517,32 @@ cdef class _ServerMiddlewareWrapper(ServerMiddleware):
...
+            CStatus status
+        if server == NULL:
+            return
+        with nogil:
+            status = server.Shutdown()
+        check_flight_status(status)
+        with nogil:
+            status = server.Wait()
+        check_flight_status(status)
+        self.server.reset()
```
Comment (pitrou) [inline python/pyarrow/_flight.pyx:2543]: `reset()` won't be called if there was an error above, could this be a problem?

File: python/pyarrow/_flight.pyx [hunk 2]
```diff
@@ -2516,6 +2517,32 @@ cdef class _ServerMiddlewareWrapper(ServerMiddleware):
...
+            PyFlightServer* server = self.server.get()
+            CStatus status
+        if server == NULL:
+            return
+        with nogil:
+            status = server.Shutdown()
+        check_flight_status(status)
+        with nogil:
+            status = server.Wait()
+        check_flight_status(status)
```
Comment (pitrou) [inline python/pyarrow/_flight.pyx:2542]: Is it instead possible to write this?
```suggestion
        with nogil:
            status = server.Shutdown() & server.Wait()
        check_flight_status(status)
```

File: python/pyarrow/_flight.pyx [hunk 3]
```diff
@@ -2550,11 +2577,13 @@ cdef class FlightServerBase(_Weakrefable):
     """
 
     cdef:
-        unique_ptr[PyFlightServer] server
+        shared_ptr[PyFlightServer] server
+        object _finalizer
```
Comment (pitrou) [inline python/pyarrow/_flight.pyx:2581]: Nit, but I don't think `cdef` members are exposed in Python, so could name it `finalizer`?

File: python/pyarrow/tests/arrow_16597.py [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+import pyarrow.flight as flight
+
+
+class Server(flight.FlightServerBase):
+    def do_put(self, context, descriptor, reader, writer):
+        time.sleep(1)
+        raise flight.FlightCancelledError("")
+
+
+server = Server("grpc://localhost:0")
```
Comment (pitrou) [inline python/pyarrow/tests/arrow_16597.py:31]: Nit: put this under a `if __name__ == "__main__"` guard?

File: python/pyarrow/tests/test_flight.py [hunk 5]
```diff
@@ -1993,6 +1993,11 @@ def test(read_all):
         descriptor = flight.FlightDescriptor.for_command(b"echo")
         writer, reader = client.do_exchange(descriptor)
         test(reader.read_all)
+        try:
+            writer.close()
+        except Exception:
+            # Silence the Cancelled/Interrupt exception
```
Comment (pitrou) [inline python/pyarrow/tests/test_flight.py:1999]: If we only want to silence a certain class of exceptions, can the `except` clause be more specific?

**Comment by pitrou** [comment]
Hmm, also this needs rebasing.

File: python/pyarrow/_flight.pyx (hunk 1 above)
Comment (lidavidm) [reply python/pyarrow/_flight.pyx]: It shouldn't be a big deal, but I wrapped it in a try-except.

File: python/pyarrow/tests/test_flight.py (hunk 5 above)
Comment (lidavidm) [reply python/pyarrow/tests/test_flight.py]: Good point, listed out KeyboardInterrupt/FlightCancelledError

**Review by pitrou — APPROVED** [review]

**Comment by pitrou** [comment]
Thank you @lidavidm !

**Pull request closed without merging**